src/nvim/api/window.c

```diff
@@ -583,3 +583,45 @@ Object nvim_win_call(Window window, LuaRef fun, Error *err)
   try_end(err);
   return res;
 }
+
+/// Gets the current viewport of a window
+///
+/// The returned tuple contains (everything is 0-based, end-exclusive):
+/// - Top-most buffer line index
+/// - Left-most column index
+/// - Number of lines renders (height)
+/// - Number of columns rendered (width)
+///
+/// @see |api-indexing|
+///
+/// @param window Window handle, or 0 for current window.
+/// @param opts Options to the function, reserved for future use.
+/// @param[out] Error deailts, if any
+///
+/// @return (row, col, height, width) tuple of the viewport
+///         amount in this window.
+ArrayOf(Integer, 4) nvim_win_get_viewport(
+    Window window,
+    Dictionary opts,
+    Error *err)
+  FUNC_API_SINCE(8)
+{
+  Array rv = ARRAY_DICT_INIT;
+  win_T *win = find_window_by_handle(window, err);
+
+  if (win) {
+    ADD(rv, INTEGER_OBJ(win->w_topline - 1));
+    ADD(rv, INTEGER_OBJ(win->w_leftcol));
+
+    linenr_T botline = win->w_botline;
+    if (buf_is_empty(win->w_buffer)) {
+      botline = 0;
+    } else if (win->w_empty_rows > 0) {
+      botline = win->w_buffer->b_ml.ml_line_count;
+    }
+    ADD(rv, INTEGER_OBJ(botline - win->w_topline + 1));
+    ADD(rv, INTEGER_OBJ(win->w_width));
```

Comment:
This doesn't acount for the margin, pretty sure this should be `win->w_width - win_col_off(win)`.

Here's a test for it:

```lua
    it('handles non-zero margins', function()
      local get_viewport, win = prepare_win { 'a', 'b', 'c' }
      meths.win_set_option(win, 'number', true);
      eq({0, 0, 2, 3}, get_viewport())
    end)
```

Reply:
The thing is that the width will always be 5 here, as this is the width of the windwo

Reply:
Width of the window but not the viewport. Plus we already have an API function for the window width. We currently have nothing that can be used to determine the margin which we a direct use for.

Reply:
I am not sure I get what you want then, in your example, the last item is `1` which is wrong, it should be `5` here.

Reply:
It is 1 because the margin is presumably 4 wide. The window width is 5, but the viewport width is 1.

Does this help?

![Screenshot 2021-08-17 at 13 38 06](https://user-images.githubusercontent.com/7904185/129728152-502af1c1-18ae-4cba-a100-cd3e6c712594.png)

From this:
- `leftcol + viewport_width = rightcol`
- `window_width - viewport_width = margin_width`

Reply:
Updated to follow this